### ISSUE TITLE
Exception on getting NSIS exe

### ISSUE THREAD

**Issue by darkurse** [tomap/nsis-extension#2]
Hi Thomas,

I'm having some issues this week to get the task going. I haven't changed anything at all on my side and it seems that I cannot get past this task anymore.

I was wondering if you noticed something similar on your side ?

To give as much feedback as possible, I've added the debug flag :

```ps1
2018-02-14T12:04:32.1367013Z ##[debug]Evaluating condition for step: 'Nsis'
2018-02-14T12:04:32.1368576Z ##[debug]Evaluating: succeeded()
2018-02-14T12:04:32.1369332Z ##[debug]Evaluating succeeded:
2018-02-14T12:04:32.1370452Z ##[debug]=> (Boolean) True
2018-02-14T12:04:32.1371228Z ##[debug]Expanded: True
2018-02-14T12:04:32.1371866Z ##[debug]Result: True
2018-02-14T12:04:32.1372548Z ##[section]Starting: Nsis
2018-02-14T12:04:32.1383898Z ==============================================================================
2018-02-14T12:04:32.1384363Z Task         : Nsis
2018-02-14T12:04:32.1384710Z Description  : Run Nsis
2018-02-14T12:04:32.1385081Z Version      : 0.3.439
2018-02-14T12:04:32.1385453Z Author       : tomap
2018-02-14T12:04:32.1385903Z Help         : Run Nsis [More Information](https://github.com/tomap/nsis-extension)
2018-02-14T12:04:32.1386456Z ==============================================================================
2018-02-14T12:04:32.1404317Z Preparing task execution handler.
2018-02-14T12:04:32.3711237Z Executing the powershell script: D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\Nsis.ps1
2018-02-14T12:04:32.3712368Z ##[debug]PowerShellHandler.Execute - AddCommand(D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\Nsis.ps1)
2018-02-14T12:04:32.3714050Z ##[debug]PowerShellHandler.Execute - Add inputParameters
2018-02-14T12:04:32.3714954Z ##[debug]PowerShellHandler.Execute - AddParameter(scriptFile=D:\a\1\s\MyProject\installers\Installer.nsi)
2018-02-14T12:04:32.3716164Z ##[debug]PowerShellHandler.Execute - AddParameter(arguments=/DDISTDIR="D:\a\1\s\MyProject\dist\windows")
2018-02-14T12:04:32.3717041Z ##[debug]PowerShellHandler.Execute - AddParameter(justInclude=no)
2018-02-14T12:04:32.3718295Z ##[debug]PowerShellHandler.Execute - AddParameter(includeMorePlugins=yes)
2018-02-14T12:04:32.4615569Z ##[debug]PowerShellHandler.Execute - Invoke
2018-02-14T12:04:33.1121630Z scriptFile = D:\a\1\s\MyProject\installers\Installer.nsi
2018-02-14T12:04:33.1122352Z 
2018-02-14T12:04:33.1122716Z 
2018-02-14T12:04:33.1125245Z arguments = /DDISTDIR="D:\a\1\s\MyProject\dist\windows"
2018-02-14T12:04:33.1125678Z 
2018-02-14T12:04:33.1125992Z 
2018-02-14T12:04:33.1127849Z justInclude = no
2018-02-14T12:04:33.1128262Z 
2018-02-14T12:04:33.1128546Z 
2018-02-14T12:04:33.1130743Z includeMorePlugins = yes
2018-02-14T12:04:33.1130986Z 
2018-02-14T12:04:33.1131150Z 
2018-02-14T12:04:35.0735020Z ##[error]System.Management.Automation.MethodInvocationException: Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel." ---> System.Net.WebException: The request was aborted: Could not create SSL/TLS secure channel.
2018-02-14T12:04:35.0736205Z ##[debug]Processed: ##vso[task.logissue type=error;]System.Management.Automation.MethodInvocationException: Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel." ---> System.Net.WebException: The request was aborted: Could not create SSL/TLS secure channel.
2018-02-14T12:04:35.0737147Z    at System.Net.WebClient.DownloadFile(Uri address, String fileName)
2018-02-14T12:04:35.0738083Z    at CallSite.Target(Closure , CallSite , Object , Object , Object )
2018-02-14T12:04:35.0738575Z    --- End of inner exception stack trace ---
2018-02-14T12:04:35.0739236Z    at System.Management.Automation.ExceptionHandlingOps.ConvertToMethodInvocationException(Exception exception, Type typeToThrow, String methodName, Int32 numArgs, MemberInfo memberInfo)
2018-02-14T12:04:35.0739924Z    at CallSite.Target(Closure , CallSite , Object , Object , Object )
2018-02-14T12:04:35.0740501Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
2018-02-14T12:04:35.0741245Z    at System.Management.Automation.Interpreter.DynamicInstruction`4.Run(InterpretedFrame frame)
2018-02-14T12:04:35.0742926Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-02-14T12:04:35.6994530Z ##[error]System.Management.Automation.MethodInvocationException: Exception calling "ExtractToDirectory" with "2" argument(s): "Could not find file 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis.zip'." ---> System.IO.FileNotFoundException: Could not find file 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis.zip'.
2018-02-14T12:04:35.6996134Z ##[debug]Processed: ##vso[task.logissue type=error;]System.Management.Automation.MethodInvocationException: Exception calling "ExtractToDirectory" with "2" argument(s): "Could not find file 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis.zip'." ---> System.IO.FileNotFoundException: Could not find file 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis.zip'.
2018-02-14T12:04:35.6997324Z    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
2018-02-14T12:04:35.6999439Z    at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
2018-02-14T12:04:35.7001096Z    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
2018-02-14T12:04:35.7001980Z    at System.IO.Compression.ZipFile.Open(String archiveFileName, ZipArchiveMode mode, Encoding entryNameEncoding)
2018-02-14T12:04:35.7002761Z    at System.IO.Compression.ZipFile.ExtractToDirectory(String sourceArchiveFileName, String destinationDirectoryName, Encoding entryNameEncoding)
2018-02-14T12:04:35.7003938Z    at CallSite.Target(Closure , CallSite , Type , Object , Object )
2018-02-14T12:04:35.7004597Z    --- End of inner exception stack trace ---
2018-02-14T12:04:35.7005257Z    at System.Management.Automation.ExceptionHandlingOps.ConvertToMethodInvocationException(Exception exception, Type typeToThrow, String methodName, Int32 numArgs, MemberInfo memberInfo)
2018-02-14T12:04:35.7005924Z    at CallSite.Target(Closure , CallSite , Type , Object , Object )
2018-02-14T12:04:35.7006464Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
2018-02-14T12:04:35.7007139Z    at System.Management.Automation.Interpreter.DynamicInstruction`4.Run(InterpretedFrame frame)
2018-02-14T12:04:35.7007670Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-02-14T12:04:35.8791289Z ##[error]System.Management.Automation.ItemNotFoundException: Cannot find path 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis' because it does not exist.
2018-02-14T12:04:35.8792220Z ##[debug]Processed: ##vso[task.logissue type=error;]System.Management.Automation.ItemNotFoundException: Cannot find path 'D:\a\_tasks\nsis_796c90b1-9f8c-424a-b03b-b789bb5945d3\0.3.439\nsis' because it does not exist.
2018-02-14T12:04:35.8793479Z    at System.Management.Automation.SessionStateInternal.GetChildItems(String path, Boolean recurse, UInt32 depth, CmdletProviderContext context)
2018-02-14T12:04:35.8794473Z    at Microsoft.PowerShell.Commands.GetChildItemCommand.ProcessRecord()
2018-02-14T12:04:35.9844497Z ##[error]System.Management.Automation.RuntimeException: Cannot index into a null array.
2018-02-14T12:04:35.9845082Z ##[debug]Processed: ##vso[task.logissue type=error;]System.Management.Automation.RuntimeException: Cannot index into a null array.
2018-02-14T12:04:35.9845722Z    at CallSite.Target(Closure , CallSite , Object , Int32 )
2018-02-14T12:04:35.9846984Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
2018-02-14T12:04:35.9847674Z    at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
2018-02-14T12:04:35.9848322Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-02-14T12:04:37.1287520Z ##[error]System.IO.DirectoryNotFoundException: Could not find a part of the path 'D:\plugins\x86-ansi'.
2018-02-14T12:04:37.1288392Z ##[debug]Processed: ##vso[task.logissue type=error;]System.IO.DirectoryNotFoundException: Could not find a part of the path 'D:\plugins\x86-ansi'.
2018-02-14T12:04:37.1289120Z    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
2018-02-14T12:04:37.1290514Z    at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
2018-02-14T12:04:37.1291242Z    at System.IO.FileInfo.CopyTo(String destFileName, Boolean overwrite)
2018-02-14T12:04:37.1291950Z    at Microsoft.PowerShell.Commands.FileSystemProvider.CopyFileInfoItem(FileInfo file, String destinationPath, Boolean force, PowerShell ps)
2018-02-14T12:04:37.1293305Z    at Microsoft.PowerShell.Commands.FileSystemProvider.CopyItemLocalOrToSession(String path, String destinationPath, Boolean recurse, Boolean Force, PowerShell ps)
2018-02-14T12:04:37.1294351Z    at Microsoft.PowerShell.Commands.FileSystemProvider.CopyItem(String path, String destinationPath, Boolean recurse)
2018-02-14T12:04:37.1295319Z    at System.Management.Automation.SessionStateInternal.CopyItem(CmdletProvider providerInstance, String path, String copyPath, Boolean recurse, CmdletProviderContext context)
2018-02-14T12:04:37.1352189Z Time taken (DL + unzip): 3 second(s)
2018-02-14T12:04:37.1471600Z ##[debug]Processed: ##vso[task.setvariable variable=NSIS_EXE;]\makensis.exe
2018-02-14T12:04:37.1472159Z 
2018-02-14T12:04:37.1472506Z 
2018-02-14T12:04:37.1499132Z Executing nsis \makensis.exe with args: /DDISTDIR="D:\a\1\s\MyProject\dist\windows" "D:\a\1\s\MyProject\installers\Installer.nsi"
[...]
2018-02-14T12:04:39.1255771Z ##[debug]Processed: ##vso[task.setprogress value=-1;] 
2018-02-14T12:04:39.1481821Z \makensis.exe /DDISTDIR="D:\a\1\s\MyProject\dist\windows" "D:\a\1\s\MyProject\installers\Installer.nsi"
2018-02-14T12:04:39.2022532Z ##[debug]Entering ProcessWrapper.CreateStartInfo
2018-02-14T12:04:39.2023295Z ##[debug]   workingFolder = 
2018-02-14T12:04:39.2024848Z ##[debug]   filename = \makensis.exe
2018-02-14T12:04:39.2026451Z ##[debug]   arguments = /DDISTDIR="D:\a\1\s\MyProject\dist\windows" "D:\a\1\s\MyProject\installers\Installer.nsi"
2018-02-14T12:04:39.2046694Z ##[debug]Leaving ProcessWrapper.CreateStartInfo
2018-02-14T12:04:39.2047686Z ##[debug]ProcessWrapper.Start - enter
2018-02-14T12:04:39.2048419Z ##[debug]ProcessWrapper.Start - starting process
2018-02-14T12:04:39.4996214Z ##[error]System.IO.FileNotFoundException: File not found: \makensis.exe
2018-02-14T12:04:39.4996799Z ##[debug]Processed: ##vso[task.logissue type=error;]System.IO.FileNotFoundException: File not found: \makensis.exe
2018-02-14T12:04:39.4997426Z File name: '\makensis.exe' ---> System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
2018-02-14T12:04:39.4998017Z    at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
2018-02-14T12:04:39.4998514Z    at Microsoft.TeamFoundation.DistributedTask.Task.LegacySDK.ProcessWrapper.Start()
2018-02-14T12:04:39.4999730Z    at Microsoft.TeamFoundation.DistributedTask.Task.LegacySDK.ProcessWrapper.Start()
2018-02-14T12:04:39.5000361Z    at Microsoft.TeamFoundation.DistributedTask.Task.LegacySDK.ProcessInvoker.Run(CancellationToken token)
2018-02-14T12:04:39.5000962Z    at Microsoft.TeamFoundation.DistributedTask.Task.Internal.InvokeToolCmdlet.ProcessRecord()
2018-02-14T12:04:39.5001516Z    at System.Management.Automation.CommandProcessor.ProcessRecord()
2018-02-14T12:04:39.5197460Z ##[error]PowerShell script completed with 6 errors.
2018-02-14T12:04:39.5199649Z ##[section]Finishing: Nsis
```

**Comment by darkurse** [comment]
Humm .. Look at this [thread on StackOverflow](https://stackoverflow.com/questions/41618766/powershell-invoke-webrequest-fails-with-ssl-tls-secure-channel) 

Could that be a solution ?

**Comment by tomap** [comment]
Hi, 
I will change this in the extension.
```
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
```
I tested it successfully

Also, I noticed that the zip is less than 3 Mb, so I believe I'll drop the download and embed nsis inside the extension
https://sourceforge.net/projects/nsis/files/NSIS%203/3.03/ 
I did originaly did that in soap ui extension because it's more than then 20 Mb extension limit https://github.com/tomap/soapui-extensions

Also, there is a new version of nsis 3.03 which does not seem to change much, So I might use this newer version instead (currently, it's 3.01)

For those last two changes, I might create a version 1.x, giving the choice in VSTS to use either the older version of the newer one

:)

**Comment by tomap** [comment]
Fixed with version 0.3.449
